src/compile/data.ts

```diff
@@ -479,12 +506,13 @@ export namespace nullFilter {
   }
 
   export function parseLayer(model: LayerModel) {
+    // note that we run this before source.parseLayer
     let nullFilterComponent = parse(model);
 
     model.children().forEach((child) => {
       const childDataComponent = child.component.data;
-      if (!childDataComponent.source && childDataComponent.nullFilter) {
-        extend(nullFilterComponent || {}, childDataComponent.nullFilter);
+      if (model.compatibleSource(child) && !differ(childDataComponent.nullFilter, nullFilterComponent)) {
```

Comment:
If two children of the layer have the same `nullFilterComponent` they still don't get merge here if I read this correctly.
Because each of the child won't be equal to the layer's initial `nullFilterComponent`.